File: cwd.go
package main

import (
	"fmt"
	"os"
	"strings"
)

func darken_prefix(prefix string) string {
	color := darken(COLOR1, CWD_DARKEN_FACTOR)

	return fmt.Sprintf("%s%s%s", fg(color), prefix, RESET)
}

// highlight only the last part of the cwd
// nothing is highlighted if the cwd is / or ~
func highlight_last(cwd string) string {
	index := max(0, strings.LastIndex(cwd, "/"))

	prefix := cwd[:index]

	if CWD_DARKEN {
		return fmt.Sprintf("%s%s%s", darken_prefix(prefix), fg(COLOR1), bold(cwd[index:]))
	}

	return fmt.Sprintf("%s%s", prefix, bold(cwd[index:]))
}

// TODO: add option to not truncate $HOME
func cwd_info() string {
	output, err := os.Getwd()

	cwd := string(output)

	if err != nil {
		return ""
	}

	output, err = os.UserHomeDir()

	home_dir := string(output)

	if err != nil {
		return cwd
	}

	if strings.HasPrefix(cwd, home_dir) {
		return "~" + strings.TrimPrefix(cwd, home_dir)
	}

	return cwd
}

func section_cwd() string {
	cwd := cwd_info()

	if CWD_HIGHLIGHT_LAST {
		cwd = highlight_last(cwd)
	} else {
		cwd = bold(cwd)
	}

	return fmt.Sprintf(
		"%s%s ",
		fg(COLOR1),
		cwd,
	)
}

File: src/cwd.rs
use dirs::home_dir;
use std::env;

use crate::config::CONFIG;
use crate::util::{bold, darken, fg};

pub fn darken_prefix(prefix: &str) -> String {
    let config = &CONFIG.modules.directory;
    let color = darken(&config.fg, CONFIG.cwd_darken_factor);

    format!("{}{}", fg(&color), prefix)
}

pub fn highlight_last(cwd: &str) -> String {
    let config = &CONFIG.modules.directory;
    let index = cwd.rfind("/").unwrap_or_default();

    let prefix = &cwd[0..index];

    if CONFIG.cwd_darken {
        format!(
            "{}{}{}",
            darken_prefix(prefix),
            fg(&config.fg),
            bold(&cwd[index..])
        )
    } else {
        format!("{}{}", prefix, bold(&cwd[index..]))
    }
}

pub fn cwd_info() -> String {
    let cwd = match env::current_dir() {
        Ok(x) => x.to_str().unwrap().to_string(),
        Err(_) => String::from("(deleted)"),
    };

    let home = match home_dir() {
        Some(x) => x.to_str().unwrap().to_string(),
        None => return cwd,
    };

    if cwd.starts_with(&home) {
        format!("~{}", &cwd[home.len()..])
    } else {
        cwd
    }
}

pub fn section_cwd() -> Option<String> {
    // NOTE: in the future, add option to specify the color of the darkened part?
    let mut cwd = cwd_info();

    if CONFIG.cwd_highlight_last {
        cwd = highlight_last(&cwd);
    } else {
        cwd = bold(&cwd);
    }

    Some(cwd)
}
